modules/animation/src/workspaceanimations.cpp

```diff
@@ -37,137 +38,228 @@ namespace animation {
-    } else if (getMainAnimationIndex() >= from) {
+    if (size() == 1) {
+        animations_[0].clear();
+        animations_[0].setName("Animation 1");
+        onChanged_.invoke(0, animations_[0]);
+        return;
+    }
+    if (size() > 1 && getMainAnimationIndex() == index) {
         // The selected one will be removed, select next after
-        auto indexBeforeRemoval = to < animations_.size() ? to : from - 1;
+        auto indexBeforeRemoval = index + 1 == animations_.size() ? index - 1 : index + 1;
         mainAnimation_.set(animations_[indexBeforeRemoval]);
-        mainAnimationIdx_ = static_cast<size_t>(
-            std::clamp(static_cast<int>(getMainAnimationIndex()) - static_cast<int>(numToRemove), 0,
-                       static_cast<int>(animations_.size()) - 1));
     }
+    Animation anim(std::move(animations_[index]));
+    animations_.erase(animations_.begin() + index);
 
-    animations_.erase(animations_.begin() + from, animations_.begin() + to);
-    names_.erase(names_.begin() + from, names_.begin() + to);
-    onChanged_.invoke(from, to - 1);
+    onChanged_.invoke(index, anim);
 }
 
-void WorkspaceAnimations::setName(size_t index, std::string_view newName) {
-    if (index < names_.size()) {
-        names_[index] = newName;
-        onChanged_.invoke(index, index);
+void WorkspaceAnimations::clear() {
+    auto nAnimations = static_cast<int>(size()) - 1;
+    for (auto i = nAnimations; i >= 0; i--) {
+        erase(i);
     }
 }
 
-void WorkspaceAnimations::setMainAnimationIndex(size_t index) {
-    Animation& anim = animations_.at(mainAnimationIdx_);
-    mainAnimationIdx_ = index;
-    mainAnimation_.set(anim);
+void WorkspaceAnimations::setMainAnimation(Animation& anim) {
+    if (auto it = find(&anim); it != end()) {
+        mainAnimation_.set(anim);
+    } else {
+        add(anim);
+        mainAnimation_.set(animations_.back());
+    }
 }
 
-size_t WorkspaceAnimations::getMainAnimationIndex() const { return mainAnimationIdx_; }
+size_t WorkspaceAnimations::getMainAnimationIndex() const {
+    // Note: cannot use WorkspaceAnimations::find here due to const, simply do the same.
+    auto it =
+        std::find_if(animations_.begin(), animations_.end(),
+                     [anim = &getMainAnimation().get()](const auto& a) { return anim == &a; });
+    return std::distance(begin(), it);
+}
 
 MainAnimation& WorkspaceAnimations::getMainAnimation() { return mainAnimation_; }
 
 const MainAnimation& WorkspaceAnimations::getMainAnimation() const { return mainAnimation_; }
 
+void WorkspaceAnimations::onAnimationChanged(AnimationController*, Animation*, Animation* newAnim) {
+    if (find(newAnim) == end()) {
+        // WorkspaceAnimations must own the Animation
+        throw Exception("MainAnimation must be set through WorkspaceAnimations");
+    }
+    // Enable undo/redo of currently selected Animation
+    app_->getProcessorNetwork()->notifyObserversProcessorNetworkChanged();
```

Comment:
humm I don't understand why we have to trigger this notification here?

Reply:
ahh this it to make the undo state dirty so it will save the state... ic... I guess we don't have a proper way of doing that, we have not enabled it for anything else before.

Reply:
ideally this should only observe animations, and the animaition should observe the track and so on. observing the multiple "levels" is a bit of a code smell also I think...

Reply:
One option would be to move the undo/redo state tracking to the UI. I could not see a good way to mark a state change though

Reply:
I guess that would involve partially moving the UndoManager to core?

Reply:
I think the workspace manager in core could be notified when something changes and also notify the undo manager for example

Reply:
Be we would need to extend the current api of course

Reply:
> I think the workspace manager in core could be notified when something changes and also notify the undo manager for example

Sounds reasonable.